Frameworks/Foundation/NSData.mm

```diff
@@ -303,69 +305,59 @@ - (instancetype)initWithContentsOfFile:(NSString*)filename {
  @Notes atomically parameter not supported
 */
 - (BOOL)writeToFile:(NSString*)filename atomically:(BOOL)atomically {
-    char* fname = (char*)[filename UTF8String];
-
-    TraceVerbose(TAG, L"NSData writing %hs (%d bytes)", fname, [self length]);
-    if (!fname) {
-        TraceVerbose(TAG, L"Filename is null!");
-        return FALSE;
-    }
-    EbrFile* fpOut = EbrFopen((const char*)fname, "wb");
-    if (fpOut) {
-        EbrFwrite([self bytes], 1, [self length], fpOut);
-        EbrFclose(fpOut);
-
-        return TRUE;
-    } else {
-        TraceVerbose(TAG, L"NSData couldn't open %hs for write", fname);
-        return FALSE;
-    }
+    return [self writeToURL:[NSURL fileURLWithPath:filename] atomically:atomically];
 }
 
 /**
  @Status Caveat
  @Notes Only file:// URLs supported. atomically parameter not supported.
 */
 - (BOOL)writeToURL:(NSURL*)url atomically:(BOOL)atomically {
-    if (![url isFileURL]) {
-        TraceVerbose(TAG, L"-[NSData::writeToURL]: Only file: URLs are supported. (%hs)", [[url absoluteString] UTF8String]);
-        return NO;
-    }
-    return [self writeToFile:[url path] atomically:atomically];
+    return [self writeToURL:url options:(atomically ? NSDataWritingAtomic : 0) error:nil];
 }
 
 /**
  @Status Caveat
  @Notes options parameter not supported
 */
 - (BOOL)writeToFile:(NSString*)filename options:(NSDataWritingOptions)options error:(NSError**)error {
-    char* fname = (char*)[filename UTF8String];
-
-    TraceVerbose(TAG, L"NSData writing %hs (%d bytes)", fname, [self length]);
-    EbrFile* fpOut = EbrFopen((const char*)fname, "wb");
-    if (fpOut) {
-        EbrFwrite([self bytes], 1, [self length], fpOut);
-        EbrFclose(fpOut);
-        return YES;
-    }
-
-    TraceVerbose(TAG, L"NSData couldn't open %hs for write (with options)", fname);
-    if (error) {
-        *error = [NSError errorWithDomain:NSCocoaErrorDomain code:NSFileNoSuchFileError userInfo:@{ NSFilePathErrorKey : filename }];
-    }
-    return NO;
+    return [self writeToURL:[NSURL fileURLWithPath:filename] options:options error:error];
 }
 
 /**
  @Status Caveat
  @Notes options parameter not supported
 */
 - (BOOL)writeToURL:(NSURL*)url options:(NSDataWritingOptions)options error:(NSError**)errorp {
-    if (![url isFileURL]) {
-        TraceVerbose(TAG, L"-[NSData::writeToURL]: Only file: URLs are supported. (%hs)", [[url absoluteString] UTF8String]);
+    if (!url) {
+        if (errorp) {
+            *errorp = [NSError errorWithDomain:@"NSData" code:100 userInfo:nil];
```

Comment:
copied from the file one now that file forwards to URL. Not the purpose of this pr to validate our error codes.

---
In reply to: [83927171](https://github.com/Microsoft/WinObjC/pull/1175#discussion_r83927171) [](ancestors = 83927171)